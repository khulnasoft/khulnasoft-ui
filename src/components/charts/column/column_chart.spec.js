import { nextTick } from 'vue';
import { shallowMount } from '@vue/test-utils';
import {
  createMockChartInstance,
  ChartTooltipStub,
  chartTooltipStubData,
} from '~helpers/chart_stubs';
import { expectHeightAutoClasses } from '~helpers/chart_height';
import {
  mockDefaultLineData,
  mockDefaultBarData,
  mockSecondaryData,
  mockDefaultStackedBarData,
} from '../../../utils/charts/mock_data';
import Chart from '../chart/chart.vue';
import ColumnChart from './column.vue';

let mockChartInstance;

jest.mock('echarts', () => ({
  getInstanceByDom: () => mockChartInstance,
}));

describe('column chart component', () => {
  const defaultChartProps = {
    xAxisTitle: 'x axis',
    yAxisTitle: 'y axis',
    xAxisType: 'category',
    bars: mockDefaultBarData,
  };
  let wrapper;

  const chartItemClickedSpy = jest.fn();
  const findChart = () => wrapper.findComponent(Chart);
  const findChartTooltip = () => wrapper.findComponent({ ref: 'dataTooltip' });

  const emitChartCreated = () => findChart().vm.$emit('created', mockChartInstance);

  const factory = (props = defaultChartProps, options = {}) => {
    wrapper = shallowMount(ColumnChart, {
      propsData: { ...props },
      listeners: {
        chartItemClicked: chartItemClickedSpy,
      },
      ...options,
    });
  };

  describe('mounted', () => {
    beforeEach(() => {
      factory();

      mockChartInstance = createMockChartInstance();
      emitChartCreated();
    });

    it('emits "created" when onCreated is called', () => {
      expect(wrapper.emitted('created')).toHaveLength(1);
    });

    it('calls event listener when "chartItemClicked" is emitted on the Chart component', () => {
      findChart().vm.$emit('chartItemClicked');

      expect(chartItemClickedSpy).toHaveBeenCalled();
    });

    it('should correctly render the chart', () => {
      const chart = findChart();

      expect(chart.props('options')).toMatchSnapshot();
    });
  });

  describe('with line data provided', () => {
    beforeEach(() => {
      factory({
        ...defaultChartProps,
        bars: [],
        lines: mockDefaultLineData,
      });
    });
    it('should correctly render the chart', () => {
      expect(findChart().props('options')).toMatchSnapshot();
    });
  });

  describe('with secondary axis data provided', () => {
    const secondaryDataTitle = 'Column test secondary';

    beforeEach(() => {
      factory({
        ...defaultChartProps,
        secondaryData: mockSecondaryData,
        secondaryDataTitle,
      });
    });
    it('should correctly render the chart', () => {
      const chart = findChart();

      expect(chart.props('options')).toMatchSnapshot();
    });

    it('should set the secondary Y axis name', () => {
      const chart = findChart();

      expect(chart.props('options').yAxis[1].name).toEqual(secondaryDataTitle);
    });
  });

  describe('with multiple series data provided', () => {
    it('should pass through custom stack series property', () => {
      factory({
        ...defaultChartProps,
        bars: mockDefaultStackedBarData.map((series) => ({
          ...series,
          stack: 'some-custom-stack',
        })),
      });

      const hasExpectedStackProperty = findChart()
        .props('options')
        .series.every((series) => series.stack === 'some-custom-stack');

      expect(hasExpectedStackProperty).toBe(true);
    });
  });

  describe('tooltip', () => {
    it('configures chart tooltip', async () => {
      factory(defaultChartProps, {
        stubs: {
          'chart-tooltip': ChartTooltipStub,
        },
      });
      mockChartInstance = createMockChartInstance();
      emitChartCreated();
      await nextTick();

      expect(findChartTooltip().props()).toMatchObject({
        chart: mockChartInstance,
        useDefaultTooltipFormatter: true,
      });
    });

    describe('custom tooltip slots', () => {
      const { params, title, content } = chartTooltipStubData;

      it('customizes value', async () => {
        const tooltipValueSlot = jest.fn().mockReturnValue('Custom tooltip value');

        factory(defaultChartProps, {
          stubs: {
            'chart-tooltip': ChartTooltipStub,
          },
          scopedSlots: {
            'tooltip-value': tooltipValueSlot,
          },
        });

        mockChartInstance = createMockChartInstance();
        emitChartCreated();
        await nextTick();

        expect(tooltipValueSlot).toHaveBeenCalledWith({ value: 9 });
        expect(findChartTooltip().text()).toBe('Custom tooltip value');
      });

      it('customizes title', async () => {
        const tooltipTitleSlot = jest.fn().mockReturnValue('Custom tooltip title');

        factory(defaultChartProps, {
          stubs: {
            'chart-tooltip': ChartTooltipStub,
          },
          scopedSlots: {
            'tooltip-title': tooltipTitleSlot,
          },
        });

        mockChartInstance = createMockChartInstance();
        emitChartCreated();
        await nextTick();

        expect(tooltipTitleSlot).toHaveBeenCalledWith({
          params,
          title,
        });
        expect(findChartTooltip().text()).toBe('Custom tooltip title');
      });

      it('customizes content', async () => {
        const tooltipContentSlot = jest.fn().mockReturnValue('Custom tooltip content');

        factory(defaultChartProps, {
          stubs: {
            'chart-tooltip': ChartTooltipStub,
          },
          scopedSlots: {
            'tooltip-content': tooltipContentSlot,
          },
        });

        mockChartInstance = createMockChartInstance();
        emitChartCreated();
        await nextTick();

        expect(tooltipContentSlot).toHaveBeenCalledWith({
          params,
          content,
        });
        expect(findChartTooltip().text()).toBe('Custom tooltip content');
      });
    });
  });

  describe('height', () => {
    expectHeightAutoClasses({
      createComponent: (props) => factory({ ...defaultChartProps, ...props }),
      findContainer: () => wrapper,
      findChart,
    });
  });
});
