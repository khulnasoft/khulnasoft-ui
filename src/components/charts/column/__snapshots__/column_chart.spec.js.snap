// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`column chart component mounted should correctly render the chart 1`] = `
Object {
  "grid": Object {
    "bottom": 44,
    "left": 64,
    "right": 32,
    "top": 16,
  },
  "legend": Object {
    "show": false,
  },
  "series": Array [
    Object {
      "barMaxWidth": "50%",
      "data": Array [
        Array [
          "Joe",
          1220,
        ],
        Array [
          "Sarah",
          932,
        ],
        Array [
          "Tom",
          901,
        ],
        Array [
          "Mary",
          934,
        ],
        Array [
          "Mike",
          1290,
        ],
        Array [
          "Ben",
          1330,
        ],
        Array [
          "Jane",
          1320,
        ],
      ],
      "emphasis": Object {
        "itemStyle": Object {
          "color": "rgba(97, 122, 226, 0.4)",
        },
      },
      "itemStyle": Object {
        "borderColor": "#617ae2",
        "borderWidth": 1,
        "color": "rgba(97, 122, 226, 0.2)",
      },
      "name": "Full",
      "stack": "stacked",
      "type": "bar",
      "yAxisIndex": 0,
    },
  ],
  "xAxis": Object {
    "axisLabel": Object {
      "margin": 20,
      "verticalAlign": "bottom",
    },
    "axisLine": Object {
      "show": false,
    },
    "axisPointer": Object {
      "type": "none",
    },
    "boundaryGap": true,
    "name": "x axis",
    "nameLocation": "center",
    "splitLine": Object {
      "show": false,
    },
    "type": "category",
  },
  "yAxis": Object {
    "axisLabel": Object {
      "formatter": [Function],
    },
    "axisTick": Object {
      "show": false,
    },
    "name": "y axis",
    "nameGap": 50,
    "nameLocation": "center",
    "type": "value",
  },
}
`;

exports[`column chart component with line data provided should correctly render the chart 1`] = `
Object {
  "grid": Object {
    "bottom": 44,
    "left": 64,
    "right": 32,
    "top": 16,
  },
  "legend": Object {
    "show": false,
  },
  "series": Array [
    Object {
      "data": Array [
        Array [
          "Joe",
          1120,
        ],
        Array [
          "Sarah",
          732,
        ],
        Array [
          "Tom",
          501,
        ],
        Array [
          "Mary",
          634,
        ],
        Array [
          "Mike",
          290,
        ],
        Array [
          "Ben",
          130,
        ],
        Array [
          "Jane",
          120,
        ],
      ],
      "itemStyle": Object {
        "color": "#617ae2",
      },
      "lineStyle": Object {
        "color": "#617ae2",
      },
      "name": "Median line",
      "type": "line",
      "yAxisIndex": 0,
    },
  ],
  "xAxis": Object {
    "axisLabel": Object {
      "margin": 20,
      "verticalAlign": "bottom",
    },
    "axisLine": Object {
      "show": false,
    },
    "axisPointer": Object {
      "type": "none",
    },
    "boundaryGap": true,
    "name": "x axis",
    "nameLocation": "center",
    "splitLine": Object {
      "show": false,
    },
    "type": "category",
  },
  "yAxis": Object {
    "axisLabel": Object {
      "formatter": [Function],
    },
    "axisTick": Object {
      "show": false,
    },
    "name": "y axis",
    "nameGap": 50,
    "nameLocation": "center",
    "type": "value",
  },
}
`;

exports[`column chart component with secondary axis data provided should correctly render the chart 1`] = `
Object {
  "grid": Object {
    "bottom": 44,
    "left": 64,
    "right": 64,
    "top": 16,
  },
  "legend": Object {
    "show": false,
  },
  "series": Array [
    Object {
      "barMaxWidth": "50%",
      "data": Array [
        Array [
          "Joe",
          1220,
        ],
        Array [
          "Sarah",
          932,
        ],
        Array [
          "Tom",
          901,
        ],
        Array [
          "Mary",
          934,
        ],
        Array [
          "Mike",
          1290,
        ],
        Array [
          "Ben",
          1330,
        ],
        Array [
          "Jane",
          1320,
        ],
      ],
      "emphasis": Object {
        "itemStyle": Object {
          "color": "rgba(97, 122, 226, 0.4)",
        },
      },
      "itemStyle": Object {
        "borderColor": "#617ae2",
        "borderWidth": 1,
        "color": "rgba(97, 122, 226, 0.2)",
      },
      "name": "Full",
      "stack": "stacked",
      "type": "bar",
      "yAxisIndex": 0,
    },
    Object {
      "barMaxWidth": "50%",
      "data": Array [
        13,
        25,
        38,
        23,
        27,
        28,
        29,
        27,
        1,
        15,
        65,
        31,
      ],
      "emphasis": Object {
        "itemStyle": Object {
          "color": "rgba(177, 79, 24, 0.4)",
        },
      },
      "itemStyle": Object {
        "borderColor": "#b14f18",
        "borderWidth": 1,
        "color": "rgba(177, 79, 24, 0.2)",
      },
      "name": "Secondary 1",
      "stack": "tiled",
      "type": "bar",
      "yAxisIndex": 1,
    },
  ],
  "xAxis": Object {
    "axisLabel": Object {
      "margin": 20,
      "verticalAlign": "bottom",
    },
    "axisLine": Object {
      "show": false,
    },
    "axisPointer": Object {
      "type": "none",
    },
    "boundaryGap": true,
    "name": "x axis",
    "nameLocation": "center",
    "splitLine": Object {
      "show": false,
    },
    "type": "category",
  },
  "yAxis": Array [
    Object {
      "axisLabel": Object {
        "formatter": [Function],
      },
      "axisTick": Object {
        "show": false,
      },
      "name": "y axis",
      "nameGap": 50,
      "nameLocation": "center",
      "type": "value",
    },
    Object {
      "axisLabel": Object {
        "formatter": [Function],
      },
      "axisTick": Object {
        "show": false,
      },
      "name": "Column test secondary",
      "nameGap": 50,
      "nameLocation": "center",
      "show": true,
      "type": "value",
    },
  ],
}
`;
